### DISCUSSION TITLE
initialData with placeholderData?

### DISCUSSION THREAD

**Discussion by J4v4Scr1pt** [TanStack/query#8251] [Q&A] [answered]
Hello!

I'm building a page where we have a table. The application is built in NextJs 14 and we initially fetch the data on the server and provide the data to the client component where we initiate the useQuery initialData with the data fetched on the server.
Then we want to keep the previous data when the user step to the next page of the table, then we use the placeholderData to do that.

But what happens now is that every time you press next page on the pagination stepper the table shows the initalData as a "placeholder" on every page.

We have tried using only placeholderData, but same behavior.

What are we doing wrong here?

```
	const { data, isLoading, isFetching, refetch } = useQuery<Organizations>({
		queryKey: [createQueryKey(filterState)],
		queryFn: async () => {
			const request = createRequest(filterState);
			try {
				return await postOrganizationFilters(request, groups, locale);
			} catch (error) {
				console.error('Error fetching Organizations', error);
				throw error;
			}
		},
		initialData: () => initialData,
		placeholderData: keepPreviousData,
		refetchOnWindowFocus: false,
		refetchOnMount: false,
	});  
```

**Comment by TkDodo** [comment]
when the key changes (`filterState`), you get a new cache entry. That cache entry then gets `initialData` passed. `initialData` puts data in the cache, and once the cache has data, `placeholderData` won’t do anything because there is no “place to hold” for it :)

you have to be specific with `initialData` when you want to apply it. I guess you only want to apply it for a specific `filterState`:

```
initialData: () => isMyFilterTheFilterThatShouldGetInitialData(filterState) ? initialData : undefined
```

**Reply by J4v4Scr1pt** [reply]
Thanks for the answer!
Ok I understand, I must have misunderstood the initialData then. When I read the docs I thought when I used it as a function it would be initiated only once when the query is initialized. But this method is actually run every time the cache entry changes?

How would you recommend working with case like this?
This is the scenario, we have Nextjs 14 App router application.
We get the initial data on the server(RSC) with default filtering applied, we fill the table with this data to have a fast load to the user. because most users don't change the filters.
Then the user can change the filters and step back and forth in the table using pagination. We want the previous page to hold the data (placeholderData) until the new data is loaded.

We have now disabled initalData because of the buggy result with it active. The page is now loaded with no table data and useQuery do the classic fetch to get initalData from on client in other words no using the power of the server and RSC. This is not 2024 way of doing it I would say 😅. 

Have been looking into TanStackTable to see if it would be a better choice for us, but I'm unsure because I have never used it 😄.

**Reply by TkDodo** [reply, marked as answer]
> When I read the docs I thought when I used it as a function it would be initiated only once when the query is initialized

that is what’s happening. but the queryKeys `['organizations', { filters: 'foo' }]` and  `['organizations', { filters: 'bar' }]` are two different queries. So when you switch filters, you might _create_ a new query if it does not exist, and then the `initialData` function runs.

> We get the initial data on the server(RSC) with default filtering applied

The options are:

- only set `initialData` when the filter equals the default filter
- use `<HydationBoundary>` instead of `initialData` for SSR. This is very likely the better choice.

**Reply by J4v4Scr1pt** [reply]
Thanks, I'm very thankful for your help, I will test with HydationBoundary.
Just providing initalData works very good in most cases, but for a table with filters this may not be the case 😅.

You are a bit bias I know... 😄, but do you think TanStackTable would fit better for this use case?

**Reply by TkDodo** [reply]
This shouldn’t matter for the query integration.

**Reply by J4v4Scr1pt** [reply]
Sorry for taking so long to get back to you.
I have followed [this ](https://tanstack.com/query/latest/docs/framework/react/guides/advanced-ssr#prefetching-and-dehydrating-data)guide in the docs. But I get the error:` No QueryClient set, use QueryClientProvider to set one`. I have searched and as I understand it it usually happens when you have several instances of react-query. But in my case this not true.

Package.json:
```
"@tanstack/react-query": "^5.61.5",
"next": "14.2.18",
"react": "18.3.1",
"react-dom": "18.3.1", 
```

I wrap my whole application in the Provider from the same page in the docs above, changed nothing.
And basically copy/paste the code from the hydration example also, only replacing my get methods.

Page.tsx:
```
import dynamic from 'next/dynamic';
import { ActivitesInitial } from './types';
import { Card, CardBody, CardHeader } from '@nextui-org/card';
import { Header } from './_Components/Header';
import { GET } from '@/utils/Api';
import { Suspense } from 'react';
import { SkeletonActivites } from './_Components/SkeletonActivites';
import { dehydrate, HydrationBoundary, QueryClient } from '@tanstack/react-query';

const Activities= dynamic(() =>
	import('./_Components/Activities').then((x) => x.Activities)
);

export default async function RecentActivitiesPage() {
	const ActivitiesResponse = (await GET(
		'api/activities?selectedGroup=undefined'
	)) as ActivitesInitial;
        const ActivitiesOnly = ActivitiesResponse?.Activities;
        
	const queryClient = new QueryClient();
	await queryClient.prefetchQuery({
		queryKey: [
			'activities',
			ActivitiesResponse.SelectedGroup,
		],
		queryFn: async () => ActivitiesOnly ,
	});

	return (
		<Card shadow="none">
			<CardHeader className="flex items-center justify-between">
				<Header
					groups={ActivitiesResponse.Groups}
					defaultSelectedGroup={ActivitiesResponse.SelectedGroup}
				/>
			</CardHeader>
			<CardBody className="flex flex-col gap-4">
				<HydrationBoundary state={dehydrate(queryClient)}>
					<Suspense fallback={<SkeletonActivites />}>
						<Activities
							initialSelectedGroup={RecentActivitiesResponse.SelectedGroup}
						/>
					</Suspense>
				</HydrationBoundary>
			</CardBody>
		</Card>
	);
}

```

Activities.tsx:
```
'use client';

import dynamic from 'next/dynamic';
import { useQuery } from '@tanstack/react-query';
import { Activity } from '../types';
import { Activity } from './Activity';
import { useQueryState } from 'nuqs';

const SkeletonActivites = dynamic(() =>
	import('./SkeletonActivites ').then((x) => x.SkeletonActivites )
);

type ActivitesProps = {
	initialSelectedGroup: string;
};

export function RecentActivities({ initialSelectedGroup }: ActivitesProps ) {
	const [selectedGroup] = useQueryState('selectedReGroup', { defaultValue: initialSelectedGroup });

	const { data: Activities, isFetching } = useQuery<Activity[]>({
		queryKey: ['activities', selectedGroup],
		queryFn: async () => {
			const response = await fetch(
				`Home/api/activites?selectedGroup=${selectedGroup}`
			);
			if (!response.ok) {
				return [];
			}
			return await response.json();
		},
		refetchOnWindowFocus: false,
	});

	if (isFetching) return <SkeletonActivites />;

	return Activities?.map((activity, index) => (
		<Activity key={activity.ActivityId + index} activity={activity} />
	));
}

```
TanstackProvider.tsx:
```

'use client';

import { isServer, QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactQueryDevtools } from '@tanstack/react-query-devtools';

function makeQueryClient() {
	return new QueryClient({
		defaultOptions: {
			queries: {
				// With SSR, we usually want to set some default staleTime
				// above 0 to avoid refetching immediately on the client
				staleTime: 60 * 1000,
			},
		},
	});
}

let browserQueryClient: QueryClient | undefined = undefined;

function getQueryClient() {
	if (isServer) {
		// Server: always make a new query client
		return makeQueryClient();
	} else {
		// Browser: make a new query client if we don't already have one
		// This is very important, so we don't re-make a new client if React
		// suspends during the initial render. This may not be needed if we
		// have a suspense boundary BELOW the creation of the query client
		if (!browserQueryClient) browserQueryClient = makeQueryClient();
		return browserQueryClient;
	}
}

export function TanstackProvider({ children }: { children: React.ReactNode }) {
	// NOTE: Avoid useState when initializing the query client if you don't
	//       have a suspense boundary between this and the code that may
	//       suspend because React will throw away the client on the initial
	//       render if it suspends and there is no boundary
	const queryClient = getQueryClient();
	return (
		<QueryClientProvider client={queryClient}>
			{children}
			<ReactQueryDevtools initialIsOpen={false} buttonPosition="bottom-right" position="bottom" />
		</QueryClientProvider>
	);
}

```

**Reply by J4v4Scr1pt** [reply]
I found the issue.
I needed to extract the HydrationBoundary to it's own component with 'use client' above like this:

```
'use client';
import { DehydratedState, HydrationBoundary } from '@tanstack/react-query';
import { ReactNode } from 'react';

interface ClientHydrationWrapperProps {
	state: DehydratedState;
	children: ReactNode;
}

export default function ClientHydrationWrapper({ state, children }: ClientHydrationWrapperProps) {
	return <HydrationBoundary state={state}>{children}</HydrationBoundary>;
}

```

The docs do not state this anywhere, is this by design or a miss?
Because the docs state this:
` // HydrationBoundary is a Client Component, so hydration will happen there.`

**Reply by TkDodo** [reply]
Shouldn't be necessary

**Reply by J4v4Scr1pt** [reply]
> Shouldn't be necessary

I know, but there is no other way for me to get it working 🤔. Do you see anything with my code above? Or the packages I use?

There is no special setup in our application. 
A simple NextJs 14 App router setup. Main layout has the providers wrapper of the children. Only thing is that this is parallel route. But that is as "special" it gets.

**Reply by TkDodo** [reply]
if there’s no special setup then please try to put it into a stackblitz / codesandbox to reproduce it.

**Reply by J4v4Scr1pt** [reply]
> if there’s no special setup then please try to put it into a stackblitz / codesandbox to reproduce it.

sure will try this. Then I guess you have not come across this issue before? Or know why it could happen?

**Reply by J4v4Scr1pt** [reply]
I have created a Stackblitz now and as you say.. it should not be needed and it works in stackblitz.

I do not understand this at all, I have tried to "talk" with both ChatGPT and v0. But none of them understands the issue and suggests wrapping the HydrationBoundary in a client component.

Could you think of any reason this would not work?

This is the [stackblitz](https://stackblitz.com/edit/stackblitz-starters-nfbco4?file=app%2F(main)%2F(Dashboard)%2FTemp%2F%40Test%2F_components%2FShowProducts.tsx), I have the exact same structure of the folders. But there are more routes ofc and some other packages like zod, sonner, framer-motion etc. But none of them should interfere with Tanstack.  
In the Stackblitz I get `Internal error: Error: 'next/dynamic css' was called outside a request scope.` error which I do not get in my project.
But this should not be related to HydrationBoundary error. If I turn of SSR for the dynamic import it goes away. Or do a regular import.  But for my "real" project I still get `No QueryClient set, use QueryClientProvider to set one` with or without dynamic import.

I have also tested the HydrationBoundary on "regular" pages without Parallel routing. Still get the error.
I have tested removing all the other Providers I have to only have the Tanstack one. Still get the error.

### Edit:
Now I'm removing all code in my application except the Tanstack code and testing with and non Parallel route. Simple Test-page with nothing extra.
I then noticed a error higher up in the console that I had overseen, shown below.
Then when I press `(TanstackProvider.tsx:34:36)` I go to the line shown in the Image below.

`main-app.js?v=1733136038095:980 Warning: Cannot update a component (`HotReload`) while rendering a different component (`HydrationBoundary`). To locate the bad setState() call inside `HydrationBoundary`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render Error Component Stack  `
 at div (<anonymous>)
    at eval (chunk-5ALFRFZW.mjs:20:13)
    at div (<anonymous>)
    at eval (chunk-H4VOEXHF.mjs:19:182)
    at ActivitiesPage [Server] (<anonymous>)
    at InnerLayoutRouter (layout-router.js:243:11)
    at RedirectErrorBoundary (redirect-boundary.js:74:9)
    at RedirectBoundary (redirect-boundary.js:82:11)
    at NotFoundBoundary (not-found-boundary.js:84:11)
    at Suspense (<anonymous>)
    at LoadingBoundary (layout-router.js:349:11)
    at ErrorBoundary (error-boundary.js:160:11)
    at InnerScrollAndFocusHandler (layout-router.js:153:9)
    at ScrollAndFocusHandler (layout-router.js:228:11)
    at RenderFromTemplateContext (render-from-template-context.js:16:44)
    at OuterLayoutRouter (layout-router.js:370:11)
    at InnerLayoutRouter (layout-router.js:243:11)
    at RedirectErrorBoundary (redirect-boundary.js:74:9)
    at RedirectBoundary (redirect-boundary.js:82:11)
    at NotFoundErrorBoundary (not-found-boundary.js:76:9)
    at NotFoundBoundary (not-found-boundary.js:84:11)
    at LoadingBoundary (layout-router.js:349:11)
    at ErrorBoundaryHandler (error-boundary.js:113:9)
    at ErrorBoundary (error-boundary.js:160:11)
    at InnerScrollAndFocusHandler (layout-router.js:153:9)
    at ScrollAndFocusHandler (layout-router.js:228:11)
    at RenderFromTemplateContext (render-from-template-context.js:16:44)
    at OuterLayoutRouter (layout-router.js:370:11)
    at div (<anonymous>)
    at main (<anonymous>)
    at QueryClientProvider (QueryClientProvider.js:27:11)
    at TanstackProvider (TanstackProvider.tsx:34:36)
    at Providers [Server] (<anonymous>)
    at body (<anonymous>)
    at html (<anonymous>)
    at RootLayout [Server] (<anonymous>)
    
    
![image](https://github.com/user-attachments/assets/1d8d6976-4653-4f1a-b21b-97833463eead)

**Reply by J4v4Scr1pt** [reply]
I have found the issue!

I have updated the Stackblitz to fail with the same issue.

This was tricky!
I had added Tanstack to the optimizePackageImports config in nextConfig. This caused the issue.

```
  experimental: {
    optimizePackageImports: ['@tanstack/react-query', '....Others'],
  },
```

You may wounder why I added this?
Well was testing with trying to decrease initial js to the client and the nextjs article about optimizePackageImports suggesting this usage. Not to add Tanstack specifically, but I added some of your packages here and forgot about it.  

Now everything works great and no InitalData is needed anymore and the preFetch pattern works as expected ✌️
